src/hb-ot-metrics.cc

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright © 2018  Ebrahim Byagowi
+ *
+ *  This is part of HarfBuzz, a text shaping library.
+ *
+ * Permission is hereby granted, without written agreement and without
+ * license or royalty fees, to use, copy, modify, and distribute this
+ * software and its documentation for any purpose, provided that the
+ * above copyright notice and the following two paragraphs appear in
+ * all copies of this software.
+ *
+ * IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE TO ANY PARTY FOR
+ * DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
+ * ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN
+ * IF THE COPYRIGHT HOLDER HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ * THE COPYRIGHT HOLDER SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING,
+ * BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ * FITNESS FOR A PARTICULAR PURPOSE.  THE SOFTWARE PROVIDED HEREUNDER IS
+ * ON AN "AS IS" BASIS, AND THE COPYRIGHT HOLDER HAS NO OBLIGATION TO
+ * PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
+ */
+
+#include "hb.hh"
+
+#include "hb-ot-gasp-table.hh"
+#include "hb-ot-os2-table.hh"
+#include "hb-ot-post-table.hh"
+#include "hb-ot-hmtx-table.hh"
+#include "hb-ot-var-mvar-table.hh"
+
+#include "hb-ot-face.hh"
+
+#if 0
+static bool
+_get_gasp (hb_font_t *font, unsigned int i, hb_position_t *result,
+	   hb_ot_metrics_t metrics_tag)
+{
+  const OT::GaspRange& range = font->face->table.gasp->get_gasp_range (i);
+  if (&range == &Null (OT::GaspRange)) return false;
+  if (result) *result = font->em_scale_y (range.rangeMaxPPEM +
+					  font->face->table.MVAR->get_var (metrics_tag, nullptr, 0));
+  return true;
+}
+#endif
+
+/**
+ * hb_ot_metrics_get:
+ * @font:
+ * @metrics_tag:
+ * @position:    (out) (optional):
+ *
+ * Returns: Whether found the requested metrics
+ *
+ * Since: REPLACEME
+ **/
+hb_bool_t
+hb_ot_metrics_get (hb_font_t       *font,
+		   hb_ot_metrics_t  metrics_tag,
+		   hb_position_t   *position     /* OUT.  May be NULL. */)
+{
+  switch (metrics_tag)
+  {
+#define GET_METRIC(TABLE, ATTR, DIR) \
+  (font->face->table.TABLE->has_data () && \
+    (position && (*position = font->em_scale_##DIR (font->face->table.TABLE->ATTR + \
+						    font->face->table.MVAR->get_var (metrics_tag, nullptr, 0))), true))
```

Comment:
The more I think about it, the more I'm convinced that we want sane fallback metrics in this API even if the respective tables are not available.  So, I'm not sure if a simple macro will do it...

Reply:
Hmm, a font without OS2 is not that imaginable I believe.

Reply:
> Hmm, a font without OS2 is not that imaginable I believe.

We don't work with "that imaginable".  We must handle all cases that happen.

Reply:
I’m pretty sure I have seen fonts in the wild (most likely on macOS) without OS/2 table.

Reply:
OK. Let me know what should happen case by case.